bootstrap/terraform/main.tf

```diff
@@ -192,47 +194,330 @@ module "eks_blueprints_crossplane_addons" {
-    provider_config = "aws-provider-config"
-    provider_aws_version = "v0.40.0"
-    # to override the default irsa policy:
-    # additional_irsa_policies = ["arn:aws:iam::aws:policy/AmazonS3FullAccess"]
+
+  depends_on = [module.eks.eks_managed_node_groups]
+}
+
+resource "kubectl_manifest" "environmentconfig" {
+  yaml_body = templatefile("${path.module}/environmentconfig.yaml", {
+    awsAccountID = data.aws_caller_identity.current.account_id
+    eksOIDC      = module.eks.oidc_provider
+    vpcID        = module.vpc.vpc_id
+  })
+
+  depends_on = [module.crossplane]
+}
+
+#---------------------------------------------------------------
+# Crossplane Providers Settings
+#---------------------------------------------------------------
+locals {
+  crossplane_namespace = "crossplane-system"
+  
+  upbound_aws_provider = {
+    enable               = true #NOTE: if you only use one aws provider, only enable one
+    version              = "v0.40.0"
+    controller_config    = "upbound-aws-controller-config"
+    provider_config_name = "aws-provider-config" #this is the providerConfigName used in all the examples in this repo
+    families = [
+      "dynamodb",
+      "elasticache",
+      "iam",
+      "kms",
+      "lambda",
+      "rds",
+      "s3",
+      "sns",
+      "sqs",
+      "vpc"
+    ]
+  }
+
+  aws_provider = {
+    enable               = false #NOTE: if you only use one aws provider, only enable one
+    version              = "v0.43.1"
+    name                 = "aws-provider"
+    controller_config    = "aws-controller-config"
+    provider_config_name = "aws-provider-config" #this is the providerConfigName used in all the examples in this repo
+  }
+
+  kubernetes_provider = {
+    enable                = true
+    version               = "v0.9.0"
+    service_account       = "kubernetes-provider"
+    name                  = "kubernetes-provider"
+    controller_config     = "kubernetes-controller-config"
+    provider_config_name  = "default"
+    cluster_role          = "cluster-admin"
+  }
+
+  helm_provider = {
+    enable                = true
+    version               = "v0.15.0"
+    service_account       = "helm-provider"
+    name                  = "helm-provider"
+    controller_config     = "helm-controller-config"
+    provider_config_name  = "default"
+    cluster_role          = "cluster-admin"
+  }
+
+}
+
+#---------------------------------------------------------------
+# Crossplane Upbound AWS Provider
+#---------------------------------------------------------------
+module "upbound_irsa_aws" {
+  count = local.upbound_aws_provider.enable == true ? 1 : 0
+  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
+  version = "~> 5.30"
+
+  role_name_prefix = "${local.name}-upbound-aws-"
+  assume_role_condition_test = "StringLike"
+
+  role_policy_arns = {
+    policy = "arn:aws:iam::aws:policy/AdministratorAccess"
```

Comment:
With the provider family configuration, it's a good idea to separate these out to service specific permissions. Not needed for this implementation.

Reply:
The service accounts are automatically created and annotated with the role arn by the `ControllerConfig`. I do not see an option where I can annotate the different SA's with different roles arns.

Reply:
Yeah I think best practice would be to have that kind of things configurable. Previously all AWS services were handled by a single deployment, but that is not the case anymore. Again, not needed for this.